From foo@baz Wed Jul 10 12:02:20 PDT 2013
Date: Wed, 10 Jul 2013 12:02:20 -0700
To: Greg KH <gregkh@linuxfoundation.org>
From: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
Subject: sysfs: add support for binary attributes in groups

groups should be able to support binary attributes, just like it
supports "normal" attributes.  This lets us only handle one type of
structure, groups, throughout the driver core and subsystems, making
binary attributes a "full fledged" part of the driver model, and not
something just "tacked on".

Reported-by: Oliver Schinagl <oliver+list@schinagl.nl>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

---
 fs/sysfs/group.c      |   20 ++++++++++++++++++--
 include/linux/sysfs.h |    4 ++--
 2 files changed, 20 insertions(+), 4 deletions(-)

--- a/fs/sysfs/group.c
+++ b/fs/sysfs/group.c
@@ -20,16 +20,19 @@ static void remove_files(struct sysfs_di
 			 const struct attribute_group *grp)
 {
 	struct attribute *const* attr;
-	int i;
+	struct bin_attribute *const* bin_attr;
 
-	for (i = 0, attr = grp->attrs; *attr; i++, attr++)
+	for (attr = grp->attrs; *attr; attr++)
 		sysfs_hash_and_remove(dir_sd, NULL, (*attr)->name);
+	for (bin_attr = grp->bin_attrs; *bin_attr; bin_attr++)
+		sysfs_remove_bin_file(kobj, *bin_attr);
 }
 
 static int create_files(struct sysfs_dirent *dir_sd, struct kobject *kobj,
 			const struct attribute_group *grp, int update)
 {
 	struct attribute *const* attr;
+	struct bin_attribute *const* bin_attr;
 	int error = 0, i;
 
 	for (i = 0, attr = grp->attrs; *attr && !error; i++, attr++) {
@@ -50,8 +53,21 @@ static int create_files(struct sysfs_dir
 		if (unlikely(error))
 			break;
 	}
+	if (error) {
+		remove_files(dir_sd, kobj, grp);
+		goto exit;
+	}
+
+	for (bin_attr = grp->bin_attrs; *bin_attr; bin_attr++) {
+		if (update)
+			sysfs_remove_bin_file(kobj, *bin_attr);
+		error = sysfs_create_bin_file(kobj, *bin_attr);
+		if (error)
+			break;
+	}
 	if (error)
 		remove_files(dir_sd, kobj, grp);
+exit:
 	return error;
 }
 
--- a/include/linux/sysfs.h
+++ b/include/linux/sysfs.h
@@ -21,6 +21,7 @@
 
 struct kobject;
 struct module;
+struct bin_attribute;
 enum kobj_ns_type;
 
 struct attribute {
@@ -59,10 +60,9 @@ struct attribute_group {
 	umode_t			(*is_visible)(struct kobject *,
 					      struct attribute *, int);
 	struct attribute	**attrs;
+	struct bin_attribute	**bin_attrs;
 };
 
-
-
 /**
  * Use these macros to make defining attributes easier. See include/linux/device.h
  * for examples..
